services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_utils.py

```diff
@@ -0,0 +1,371 @@
+                msg=(
+                    "Timed out while searching for an assigned NodeID for "
+                    f"service_id={service_id}. Last task inspect result: {task}"
+                )
+            )
+
+    async with docker_client() as client:  # pylint: disable=not-async-context-manager
+        service_state: str = None
+        started = time.time()
+
+        while service_state not in target_statuses:
+            running_services = await client.tasks.list(filters={"service": service_id})
+
+            service_container_count = len(running_services)
+
+            # the service could not be started yet, let's wait for the next iteration?
+            if service_container_count == 0:
+                await sleep_or_error(started=started, task={})
+                continue
+
+            # The service might have more then one task because the previous might have died out
+            # Only interested in the latest Task/container as only 1 container per service
+            # is being run
+            sorted_tasks = sorted(running_services, key=lambda task: task["UpdatedAt"])
+            task = sorted_tasks[-1]
+
+            service_state = task["Status"]["State"]
+
+            await sleep_or_error(started=started, task=task)
+
+    return task
+
+
+async def get_node_id_from_task_for_service(
+    service_id: str, dynamic_sidecar_settings: DynamicSidecarSettings
+) -> str:
+    """Awaits until the service has a running task and returns the
+    node's ID where it is running. When in a running state, the service
+    is most certainly has a NodeID assigned"""
+
+    task = await _extract_task_data_from_service_for_state(
+        service_id=service_id,
+        dynamic_sidecar_settings=dynamic_sidecar_settings,
+        target_statuses=TASK_STATES_RUNNING,
+    )
+
+    if "NodeID" not in task:
+        raise DynamicSidecarError(
+            msg=(
+                f"Could not find an assigned NodeID for service_id={service_id}. "
+                f"Last task inspect result: {task}"
+            )
+        )
+
+    return task["NodeID"]
+
+
+async def get_dynamic_sidecar_state(
+    service_id: str, dynamic_sidecar_settings: DynamicSidecarSettings
+) -> Tuple[ServiceState, str]:
+
+    try:
+        last_task = await _extract_task_data_from_service_for_state(
+            service_id=service_id,
+            dynamic_sidecar_settings=dynamic_sidecar_settings,
+            target_statuses=TASK_STATES_ALL,
+        )
+    # GenericDockerError
+    except GenericDockerError as e:
+        if e.original_exception.message != f"service {service_id} not found":
+            raise e
+
+        # because the service is not there yet return a pending state
+        # it is looking for a service or something with no error message
+        pending_task_state = {"State": ServiceState.PENDING.value}
+        return extract_task_state(task_status=pending_task_state)
+
+    task_status = last_task["Status"]
```

Comment:
question, so if I get it well you return the state of the dynamic sidecar until it is available?
that means it returns:
- PENDING, PULLING (sidecar), STARTING (sidecar), RUNNING (sidecar) and then again PENDING, PULLING, STARTING, RUNNING
is that correct?
if yes, I think this is confusing for the user. The user needs to know about his service status. so my proposal was to return PENDING as long as the sidecar is not up and running. then return the service effective state.

Reply:
OK, changed this as requested.